src/bootstrap/src/bin/rustc.rs

```diff
@@ -95,7 +95,6 @@ fn main() {
     // When statically linking `std` into `rustc_driver`, remove `-C prefer-dynamic`
     if env::var("RUSTC_LINK_STD_INTO_RUSTC_DRIVER").unwrap() == "1"
         && crate_name == Some("rustc_driver")
-        && stage != "0"
```

Comment:
Why?

Reply:
This was added back when the beta compiler did not use static linking. Now beta rustc statically links to libstd, so this shouldn't be needed anymore, AFAIK.

Reply:
stage0 can be configured to custom compiler which uses dynamic linking.

Reply:
Right, but relatively soon most rustc compilers being used should statically link to their libstd. Do we need to support old(er) versions?

Reply:
It's not only about older versions. I just don't want to get a Zulip notification from people building with their custom stage0 compiler and running into this issue. If it's not causing to too much complexity on bootstrap, supporting as many cases as possible is always a good thing in my opinion.

Reply:
Currently bootstrap produces the statically linked rustc unconditionally, IIRC (on supported platforms), so "custom stage0 compiler" would have to be something where people modified their bootstrap code, at which point anything could happen I suppose.

Without removing this condition, the `stage1/lib` directory would still be required to contain `libstd.so`. That's both surprising, because it does not match what happens on later stages, and it could also conceal potential test failures (or tests succeeding unexpectedly), as is the test that I had to fix in this PR.

As usually, the stage numbering is confusing here - removing the condition alters what happens with the stage1 compiler.

Reply:
> Currently bootstrap produces the statically linked rustc unconditionally

If that's the case yeah this seems reasonable, but what about windows-gnu target?

https://github.com/rust-lang/rust/blob/56e35a5dbb37898433a43133dff0398f46d577b8/src/bootstrap/src/core/builder.rs#L1110-L1114

Reply:
That will not receive `RUSTC_LINK_STD_INTO_RUSTC_DRIVER`, so it will be still linked dynamically, regardless of the stage :)

Reply:
This stage0 check is to ensure support for the `rustc_private` linkage change, which is now present in bootstrap, even on `windows-gnu`.

Reply:
Removal needs to be reverted then. Can you add a comment saying that this check is used for respecting stage0's linkage mode?

Reply:
FWIW linkage mode should be handled with a config option instead of setting that unconditionally IMO, but that's another topic.

Reply:
> FWIW linkage mode should be handled with a config option instead of setting that unconditionally IMO

I think the 3 of us here would disagree on this ^^

Reply:
> > FWIW linkage mode should be handled with a config option instead of setting that unconditionally IMO
> 
> 
> 
> I think the 3 of us here would disagree on this ^^

What is it so bad about it? You can still keep the current behaviour as a default.

Reply:
I consider the previous behavior a hack and am happy to see it disappear. I don't know of any benefits to it.

Reply:
> FWIW linkage mode should be handled with a config option instead of setting that unconditionally IMO, but that's another topic.

There's AFAIK absolutely no reason why rustc should link to libstd dynamically. The libstd is tied 1:1 to librustc_driver, it can't really be replaced by a different version. Linking statically makes the compiler faster and reduces the size of the compiler sysroot, which saves us $ :D Sadly, we cannot use it everywhere yet, because of the windows problems, but that's a bug, which should be worked around with an if condition in bootstrap, rather than a config option, IMO.

Reply:
I get that we're not interested in dynamic linking and I never said we were. Adding a config option doesn't mean that we will use it in our CI or default profiles. It’s just about how we design our build system; trying to make it as configurable as possible. Is the reason this isn't configurable because we don't want anyone in the world to use static linking? If that’s the case, fine. But tbh I haven't seen any solid reason that justifies not having a config option so far.

Reply:
> Removal needs to be reverted then. Can you add a comment saying that this check is used for respecting stage0's linkage mode?

No, it was just there temporarily until the bootstrap compiler was bumped.

Reply:
> Is the reason this isn't configurable because we don't want anyone in the world to use static linking?

I would in fact argue that the anwer to `Is the reason this isn't configurable because we don't want anyone in the world to use static linking?` is "yes" (well, it should have said "dynamic" instead of "static", but that was a typo :) ). As long as static linking works for your target, I don't think that there is any reason to link to libstd dynamically. The default was swapped a few months ago and I haven't seen anyone complaining so far.

I'm not opposed to making this configurable, but I wouldn't do it proactively. If someone comes with a reasonable use-case why they need it, then sure. It's possible that there are indeed use-cases for this, but I'd like to see them first. We already have a lot of config options and I don't think we necessarily have to add new ones without a specific use-case.

Reply:
> I'm not opposed to making this configurable, but I wouldn't do it proactively. If someone comes with a reasonable use-case why they need it, then sure. It's possible that there are indeed use-cases for this, but I'd like to see them first. We already have a lot of config options and I don't think we necessarily have to add new ones without a specific use-case.

Yeah this makes more sense.